vote-sc/src/lib.rs

```diff
@@ -0,0 +1,66 @@
+#![no_std]
+
+multiversx_sc::imports!();
+
+mod caller_check;
+pub mod constants;
+mod errors;
+pub mod events;
+pub mod liquid_staking_proxy;
+pub mod views;
+
+use crate::{constants::*, errors::*};
+
+#[multiversx_sc::contract]
+pub trait VoteSC:
+    caller_check::CallerCheckModule + events::EventsModule + views::ViewsModule
+{
+    #[init]
+    fn init(&self) {}
+
+    #[only_owner]
+    #[endpoint]
+    fn set_root_hash(&self, root_hash: Hash<Self::Api>, proposal_id: ProposalId) {
+        require!(!root_hash.is_empty(), INVALID_ROOT_HASH);
+        self.root_hash_proposal_nonce(proposal_id).set(root_hash)
+    }
+
+    #[only_owner]
+    #[endpoint]
+    fn set_liquid_staking_address(&self, address: ManagedAddress) {
+        require!(
+            self.blockchain().is_smart_contract(&address),
+            INVALID_SC_ADDRESS
+        );
+        self.liquid_staking_sc().set(address);
+    }
+
+    #[endpoint]
+    fn vote(
+        &self,
+        proposal_id: ProposalId,
+        vote: ManagedBuffer,
+        voting_power: BigUint<Self::Api>,
+        proof: ArrayVec<ManagedByteArray<HASH_LENGTH>, PROOF_LENGTH>,
+    ) {
+        self.require_caller_not_self();
+
+        let voter = self.blockchain().get_caller();
+        require!(
+            self.confirm_voting_power(proposal_id, voting_power.clone(), proof),
+            INVALID_MERKLE_PROOF
+        );
+
+        require!(!self.liquid_staking_sc().is_empty(), LS_SC_NOT_SET);
```

Comment:
Being a simple check, maybe it is more efficient at the start of the function, before verifying the merkle proof.